### ISSUE TITLE
h2ph generates incorrect code for '#if defined A || defined B'

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8449]
Migrated from [rt.perl.org#39130](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=39130) (status was 'resolved')

Searchable as RT39130$

**Comment by p5pRT** [comment]
### From jvdias@redhat.com

#### Created by jvdias@redhat.com



For cpp statements like :

&nbsp;  \#if defined A || defined B

h2ph would generate the code :

&nbsp;  if\(defined \(defined\(&A\) ? &A : 0\) || defined \(defined\(&B\) ? &B : 0\)\) \{

which is tautologous\, as defined\(0\) is always true \.

This turns out to cause real problems on the Linux ppc64 platform\,
where "endian\-ness" is selectable in the compiler; every inclusion
of "endian\.ph"\, which includes "bits/endian\.ph"\, generates an error&#8203;:
 
$ perl \-e 'require "sys/socket\.ph";'
Both BIG\_ENDIAN and LITTLE\_ENDIAN defined\! at
/usr/lib/perl5/5\.8\.8/ppc\-linux\-thread\-multi/bits/endian\.ph line 10\.
Compilation failed in require at\.\.\.

Because h2ph generated this code in \.\./bits/endian\.ph&#8203;:
if\(defined \(defined\(&\_\_BIG\_ENDIAN\_\_\) ? &\_\_BIG\_ENDIAN\_\_ : 0\) || defined \(defined\(&\_BIG\_ENDIAN\) ? &\_BIG\_ENDIAN : 0\)\) \{
&nbsp;    if\(defined \(defined\(&\_\_LITTLE\_ENDIAN\_\_\) ? &\_\_LITTLE\_ENDIAN\_\_ : 0\) || defined \(defined\(&\_LITTLE\_ENDIAN\) ? &\_LITTLE\_ENDIAN : 0\)\) \{
&nbsp;        die\("Both\\ BIG\_ENDIAN\\ and\\ LITTLE\_ENDIAN\\ defined\\\!"\);
&nbsp;    \}
For the cpp code in /usr/include/bits/endian\.h&#8203;:
\#if defined \_\_BIG\_ENDIAN\_\_ || defined \_BIG\_ENDIAN
\# if defined \_\_LITTLE\_ENDIAN\_\_ || defined \_LITTLE\_ENDIAN
\#  error Both BIG\_ENDIAN and LITTLE\_ENDIAN defined\!
\.\.\.

The problem does not happen if the 'defined\.\.' conditions are in parentheses \-
ie\.
&nbsp;  \#if defined\(A\) || defined\(B\)
produces correct code&#8203;:
&nbsp;  if\( defined\(&A\) || defined\(&B\) \)

A trivial fix for this is to simply replace the '&#8203;: 0' with '&#8203;: undef' in
h2ph\.PL line 517\, as with this patch&#8203;:
<details><summary>Inline Patch</summary>

```diff
--- perl-5.8.8/utils/h2ph.PL~   2006-01-12 17:55:04.000000000 -0500
+++ perl-5.8.8/utils/h2ph.PL    2006-05-11 13:50:04.000000000 -0400
@@ -514,7 +514,7 @@
                }
            } else {
                if ($inif && $new !~ /defined\s*\($/) {
-                   $new .= '(defined(&' . $id . ') ? &' . $id . ' : 0)';
+                   $new .= '(defined(&' . $id . ') ? &' . $id . ' : undef)';
                } elsif (/^\[/) {
                    $new .= " \$$id";
                } else {



```
</details>
Please consider applying the above patch / fixing this issue in maint .

Thanks & Regards\,

Jason Vas Dias \<jvdias@&#8203;redhat\.com>
perl package maintainer
Red Hat\, Inc\.   

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

This perlbug was built using Perl v5.8.8 in the Red Hat build system.
It is being executed now by Perl v5.8.8 - Fri Apr 21 17:40:46 EDT 2006.

Site configuration information for perl v5.8.8:

Configured by Red Hat, Inc. at Fri Apr 21 17:40:46 EDT 2006.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=linux, osvers=2.6.16-1.2136_fc6, archname=i386-linux-thread-multi
    uname='linux jvdias 2.6.16-1.2136_fc6 #1 smp tue apr 18 14:56:23 edt 2006 i686 i686 i386 gnulinux '
    config_args='-des -Doptimize=-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m32 -march=i386 -mtune=generic -fasynchronous-unwind-tables -Dversion=5.8.8 -Dmyhostname=localhost -Dperladmin=root@localhost -Dcc=gcc -Dcf_by=Red Hat, Inc. -Dinstallprefix=/usr -Dprefix=/usr -Darchname=i386-linux -Dvendorprefix=/usr -Dsiteprefix=/usr -Duseshrplib -Dusethreads -Duseithreads -Duselargefiles -Dd_dosuid -Dd_semctl_semun -Di_db -Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio -Dinstallusrbinperl=n -Ubincompat5005 -Uversiononly -Dpager=/usr/bin/less -isr -Dd_gethostent_r_proto -Ud_endhostent_r_proto -Ud_sethostent_r_proto -Ud_endprotoent_r_proto -Ud_setprotoent_r_proto -Ud_endservent_r_proto -Ud_setservent_r_proto -Dinc_version_list=5.8.7 5.8.6 5.8.5 5.8.4 5.8.3 -Dscriptdir=/usr/bin'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING -fno-strict-aliasing -pipe -Wdeclaration-after-statement -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m32 -march=i386 -mtune=generic -fasynchronous-unwind-tables',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING -fno-strict-aliasing -pipe -Wdeclaration-after-statement -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='4.1.0 20060420 (Red Hat 4.1.0-9)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lresolv -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lresolv -lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=/lib/libc-2.4.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.4'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/usr/lib/perl5/5.8.8/i386-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.8:
    /usr/lib/perl5/site_perl/5.8.8/i386-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.7/i386-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.6/i386-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.5/i386-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.4/i386-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.3/i386-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.8
    /usr/lib/perl5/site_perl/5.8.7
    /usr/lib/perl5/site_perl/5.8.6
    /usr/lib/perl5/site_perl/5.8.5
    /usr/lib/perl5/site_perl/5.8.4
    /usr/lib/perl5/site_perl/5.8.3
    /usr/lib/perl5/site_perl
    /usr/lib/perl5/vendor_perl/5.8.8/i386-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.7/i386-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.6/i386-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.5/i386-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.4/i386-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.3/i386-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.8
    /usr/lib/perl5/vendor_perl/5.8.7
    /usr/lib/perl5/vendor_perl/5.8.6
    /usr/lib/perl5/vendor_perl/5.8.5
    /usr/lib/perl5/vendor_perl/5.8.4
    /usr/lib/perl5/vendor_perl/5.8.3
    /usr/lib/perl5/vendor_perl
    /usr/lib/perl5/5.8.8/i386-linux-thread-multi
    /usr/lib/perl5/5.8.8
    .


Environment for perl v5.8.8:
    HOME=/home/jvd
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/kerberos/bin:/usr/bin:/bin:/usr/bin:/usr/games:/usr/X11R6/bin:/home/jvd/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'

**Comment by p5pRT** [comment]
### From @rgs

Jason Vas Dias \(via RT\) wrote&#8203;:
> For cpp statements like :
> 
>   \#if defined A || defined B
> 
> h2ph would generate the code :
> 
>   if\(defined \(defined\(&A\) ? &A : 0\) || defined \(defined\(&B\) ? &B : 0\)\) \{
> 
> which is tautologous\, as defined\(0\) is always true \.
\.\.\.
> A trivial fix for this is to simply replace the '&#8203;: 0' with '&#8203;: undef' in
> h2ph\.PL line 517\, as with this patch&#8203;:
> \-\-\- perl\-5\.8\.8/utils/h2ph\.PL~   2006\-01\-12 17&#8203;:55&#8203;:04\.000000000 \-0500
> \+\+\+ perl\-5\.8\.8/utils/h2ph\.PL    2006\-05\-11 13&#8203;:50&#8203;:04\.000000000 \-0400
> @&#8203;@&#8203; \-514\,7 \+514\,7 @&#8203;@&#8203;
>                 \}
>             \} else \{
>                 if \($inif && $new \!~ /defined\\s\*\\\($/\) \{
> \-                   $new \.= '\(defined\(&' \. $id \. '\) ? &' \. $id \. ' : 0\)';
> \+                   $new \.= '\(defined\(&' \. $id \. '\) ? &' \. $id \. ' : undef\)';

Thanks\, applied to bleadperl as change \#28225\.

> Please consider applying the above patch / fixing this issue in maint \.

\(To Nicholas&#8203;:\) I concur\.
